### PR TITLE
Improve null_object usage: assignment/compare

### PR DESCRIPTION
**Pull request by errael**
Address #12043
Fix "xvar == null", where xvar might have been assigned null_object.
Fix compilation failure: "var o2: C = null_object".

### CODE REVIEW

**Comment by errael** [comment]
@yegappan This fixes some cases, I haven't played with it thoroughly yet and it's draft. I copied in t_object stuff from the instanceof PR; I'll update when that gets merged. It's a relatively small PR and no new tests yet. Simple fixes to COMPAREANY/COMPARENULL. And a compile fix.

But I have questions...

- `generate_PUSHOBJ` now invoked from another file. How to create proto? Manually?
- Would `t_object_null` be better than `t_object`?
- I'm not sure about the compile fix; I left a few TODO in there, comments appreciated.

I ran
```
make $(ls test_vim9*.vim | sed 's/\.vim//g'|grep -v test_vim9_builtin)
```
without errors. test_vim9_builtin got 3 errors with or without this PR.

File: src/vim9type.c
Comment (yegappan) [inline src/vim9type.c]: Instead of this check, can you use the following?
```
+           // tt_class is NULL for an "unknown" object (either a null object
+           // or an uninitialized object)
+           if (actual->tt_class == NULL)
+               return OK;
+
```

File: src/vim9type.c
Comment (yegappan) [inline src/vim9type.c]: I am working on adding the test_null_class() and test_null_object() functions.  I am using the following test.  You can leverage this test for this PR and remove the references to the test_null_class() and test_null_object() functions:

```
" Test for the test_null_object() and test_null_class() functions
def Test_null_class_object()
  var lines =<< trim END
    vim9script

    var c = test_null_class()
    var obj = test_null_object()
    assert_true(c == null_class)
    assert_true(obj == null_object)
    assert_fails('echo c != obj', 'E1072:')
    assert_equal('class<Unknown>', typename(c))
    assert_equal('object<Unknown>', typename(obj))
    assert_true(c == null)
    assert_true(obj == null)

    var nullo = null_object
    def F(): any
      return nullo
    enddef
    assert_equal('object<Unknown>', typename(F()))

    var o1: any = nullo
    assert_true(o1 == null_object)
    assert_true(o1 == null)

    def G()
      var x = null_object
    enddef

    class C
    endclass
    var o2: C
    o2 = null_object
  END
  v9.CheckScriptSuccess(lines)
enddef
```

File: src/vim9type.c
Comment (yegappan) [inline src/vim9type.c:912]: With the support for the null_object, the existing Test_object_not_set() test in test_vim9_class.vim is failing with the `E1360: Using a null object` error instead of the `E1012: Type mismatch; expected object<Background> but got object<Unknown>` error.  You need to update that test with the new error.

File: src/globals.h
Comment (yegappan) [inline src/globals.h]: I assume you will remove this comment once the instanceof() PR is merged?

File: src/globals.h
Comment (yegappan) [inline src/globals.h]: We can keep the below definitions instead of renaming this to t_object_null.

**Review by yegappan — CHANGES_REQUESTED** [review]

File: src/vim9type.c
Comment (errael) [reply src/vim9type.c]: I had it that way initially; got test failures. I didn't track it down, I'll put it back and check out the failures.

File: src/globals.h
Comment (errael) [reply src/globals.h]: Right

File: src/vim9type.c
Comment (yegappan) [reply src/vim9type.c]: With this change, the tests are passing for me (except for the one change to Test_object_not_set() mentioned in the other comment).

File: src/globals.h
Comment (errael) [reply src/globals.h]: OK. I don't really grok what these static types are all about and how they're used. So any a VAR_OBJECT with type where `tt_class == NULL` is a null object.

File: src/vim9type.c
Comment (errael) [reply src/vim9type.c]: Is this getting merged soon? These new tests go into test_vim9_class.vim?

Comment (yegappan) [reply src/vim9type.c]: The new test function is in the test_vim9_class.vim file.  You can leverage this test and modify it as needed and feel free to include this in your PR.  I will wait for your PR to be merged.

Comment (errael) [reply src/vim9type.c]: OK. And this PR needs #12867.

File: src/typval.c
Comment (yegappan) [inline src/typval.c:1581]: Is this comment still applicable?

**Review by yegappan — CHANGES_REQUESTED** [review]

File: src/typval.c
Comment (errael) [reply src/typval.c:1581]: Yes. I didn't do anything for or about a variable that aliases a class.  I accidentally found out that you can do
```
var vclass = C
var o = vclass.new()
```
and mentioned it in https://github.com/vim/vim/issues/12043#issuecomment-1688584460 where I questioned whether or not it was supposed to work. Other than that, I didn't try anything. I've since found in todo.txt
> When "Meta" is a class, is "const MetaAlias = Meta" allowed?  It should
> either work or given an error. possibly give an error now and implement it
> later (using a typedef).  #12006

So this is a pretty open question. Might be good idea to make the assignment fail until there's a plan on how to handle this.

I'm pretty sure it won't work in general. I don't think there is a `class` type. Certainly `var foo: class` doesn't work.

But, if you want `vclass == null` then the commented part is needed, along with plenty more.

**Review by yegappan — APPROVED** [review]

**Comment by errael** [comment]
Rebase to current master

**Pull request closed without merging**
